lib/uri_service/solr.rb

```diff
@@ -0,0 +1,81 @@
+module URIService
+  class Solr
+    attr_accessor :connection
+
+    # Returns configuration options set in solr.yml
+    def connection_config
+      Rails.application.config_for(:solr).symbolize_keys
+    end
+
+    # Returns RSolr connection to Solr instance
+    def connection
+      @connection ||= RSolr.connect(connection_config)
+    end
+
+    # Look up term
+    #
+    # @param String vocabulary vocabulary string key
+    # @param String uri
+    # @return nil if no matching term found
+    # @return Hash if matching term found
+    def find_term(vocabulary, uri)
+      results = search do |params|
+        params.vocabulary(vocabulary).uri(uri)
+      end
+
+      case results['response']['numFound']
+      when 0
+        nil
+      when 1
+        results['response']['docs'].first
+      else
+        raise 'More than one term document matched uri'
+      end
+    end
+
+    # Solr query. Returns solr json.
+    def search
+      search_parameters = SolrParams.new
+
+      yield(search_parameters)
+
+      params = search_parameters.to_h
+
+      # If making a search use the /search handler otherwise use /select. /select
+      # queries with just filters are faster than /search queries.
+      handler = (params[:q].blank?) ? 'select' : 'search'
+
+      connection.get(handler, params: params)
+    end
+
+    # Add document
+    #
+    # @param Hash json document to be added to solr
+    def add(doc)
+      connection.add(doc)
+      connection.commit # probably need to make this optional or leverage soft commit
```

Comment:
I was thinking more about this actually.  We don't want to commit constantly, but we do want to always make sure that our retrieved data is 100% up to day.  This leaves us with two options:

1. Always pull term data from the database.  i.e. Do a solr search (and don't store anything in the solr index other than uuids), then use that list of returned UUIDs to do *one* db query to retrieve all of the terms in that UUID list.   I'm hoping that a specific db query based on a list of UUIDs will be really fast.  If it gets slow, we can consider an intermediate document store later, which could either have a limited size or could potentially store 100% of terms if that's not unreasonable.  On 

2. Same idea as above, but introduce an intermediate non-solr caching layer for recently updated terms and draw from this cache when returning term search results, and fall back to the database when something isn't found.  (I don't love the additional complexity though.)

What do you think?

Reply:
Wouldn't searches not reflect the changes made if we don't commit them?

Reply:
I had an idea about this. Maybe we can make all filter queries database queries since we are matching by exact string. We can use solr to make request that use the `q` parameter, because that does fuzzy searching and what not. That way only queries that use the q parameter might be partially out of sync. We can talk about it later in person.